Fix `retry_exponential_backoff` divide by zero error when retry delay is zero

edwardwang888 opened the pull request:
When `retry_delay` is zero, the `retry_exponential_backoff` algorithm has a divide-by-zero error in the `modded_hash` calculation, which causes the scheduler to crash.

https://github.com/apache/airflow/commit/d1dceff69ab15f9ce28ea04215a693b81c8a93a2 attempted to address the divide-by-zero error by adding a ceiling function, but it doesn't account for this particular edge case. Using a lower bound instead would be a better implementation.

Fixes #17005.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

uranusjr commented on airflow/models/taskinstance.py:
```diff
@@ -939,9 +939,9 @@ def next_retry_datetime(self):
-            min_backoff = int(math.ceil(delay.total_seconds() * (2 ** (self.try_number - 2))))
+            # we must impose a lower bound of 1 prior to converting to an int, otherwise a divide
+            # by zero error will occur in the modded_hash calculation.
+            min_backoff = int(max(1, delay.total_seconds() * (2 ** (self.try_number - 2))))
```
This changes the value for (say) `delay=timedelta(seconds=1.5), self.try_number=2`. Previously the result would be 2 (`ceil(1.5 * 1)`) but now it’s 1 (`int(1.5 * 1)`).

edwardwang888 replied:
Yeah, actually previously it was defined without the ceiling - it looks like the ceiling was added to deal with divide-by-zero issue (https://github.com/apache/airflow/commit/d1dceff69ab15f9ce28ea04215a693b81c8a93a2). So I guess it is fine to revert to the previous definition, and just add a lower bound.

uranusjr replied:
That change was made between 1.x and 2.x though and could be categorised as a breaking change. Not sure (one second isn’t really that important either way), maybe other maintainers can chime in on this.

eladkal replied:
How about replacing the retry_delay value before the calculation?

```
modified_delay = delay.total_seconds() if delay.total_seconds() > 0 else 1
min_backoff = int(math.ceil(modified_delay * (2 ** (self.try_number - 2))))
```

example:
```
import math
import numpy as np
from datetime import timedelta

for i in np.arange(0, 3, 0.5):
    for try_number in range(1, 3, 1):
        delay = timedelta(seconds=i)

        min_backoff = int(math.ceil(delay.total_seconds() * (2 ** (try_number - 2))))

        modified_delay = delay.total_seconds() if delay.total_seconds() > 0 else 1
        min_backoff_new = int(math.ceil(modified_delay * (2 ** (try_number - 2))))

        print(f"Input(seconds: {i} , try_number: {try_number}) --> Result(Prev: {min_backoff}, New: {min_backoff_new})")

```
Output:
```
Input(seconds: 0.0 , try_number: 1) --> Result(Prev: 0, New: 1)
Input(seconds: 0.0 , try_number: 2) --> Result(Prev: 0, New: 1)
Input(seconds: 0.5 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 0.5 , try_number: 2) --> Result(Prev: 1, New: 1)
Input(seconds: 1.0 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 1.0 , try_number: 2) --> Result(Prev: 1, New: 1)
Input(seconds: 1.5 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 1.5 , try_number: 2) --> Result(Prev: 2, New: 2)
Input(seconds: 2.0 , try_number: 1) --> Result(Prev: 1, New: 1)
Input(seconds: 2.0 , try_number: 2) --> Result(Prev: 2, New: 2)
Input(seconds: 2.5 , try_number: 1) --> Result(Prev: 2, New: 2)
Input(seconds: 2.5 , try_number: 2) --> Result(Prev: 3, New: 3)
```

edwardwang888 replied:
That should work. Is the reason for using `modified_delay` to prevent this from becoming a breaking change? (I'm curious how significant the one second difference is.)

eladkal replied:
What is the breaking change?
For `0` it never worked, for the rest of values they stay the same.

edwardwang888 replied:
I meant that my original solution is a breaking change, even though the difference is at most one second. In practice, this is not a big difference, but I am curious how significant the breaking change is from a code perspective.

eladkal replied:
Personally I don't think in this case 1 second is that important.
That said, we have another solution that avoid the need to debate/raise this question so why not take it? :)

eladkal replied:
@edwardwang888 would you like to finish this PR?

edwardwang888 replied:
@eladkal Yes, sorry about the delay. I plan to finish it sometime this week - hopefully it should be a small change.

edwardwang888 replied:
@eladkal I have updated the PR. One test is failing but the failure looks unrelated to this change.

potiuk replied:
I am good with the approach as is now. @eladkal ?

eladkal replied:
Yep

edwardwang888 committed:
Add lower bound for min_backoff

edwardwang888 committed:
Fix test calculation

This test was incorrect before. When `delay.total_seconds()` equals 1 and `self.try_number` equals 1, `min_backoff` equals 1. Thus, `modded_hash` always equals 1, regardless of the value of `ti_hash`. So it should not be a period of 1 to 15 seconds, but just 1 second.

Code referenced was:
airflow/airflow/models/taskinstance.py
Lines 1064 to 1087

edwardwang888 committed:
Expand test to cover zero seconds

The test can be reused as the end result is the same.
